File: Examples/FxSocket.Sftp/Windows/CS/SftpDesktopClientDemo/TransferIssueForm.cs
using System;
using System.IO;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using FxSocket.IO;
using FxSocket.FileSystem;
using FxSocket.Net;

namespace FxSocketSamples
{
	/// <summary>
	/// Problem handling form.
	/// </summary>
	public class TransferIssueForm : System.Windows.Forms.Form
	{
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;

        private System.Windows.Forms.Label lblMessage;
		private System.Windows.Forms.RadioButton btnSkip;
        private System.Windows.Forms.RadioButton btnRename;
        private System.Windows.Forms.RadioButton btnOverwrite;
		private System.Windows.Forms.RadioButton btnOverwriteOlder;
        private System.Windows.Forms.RadioButton btnOverwriteDiffSize;
		private System.Windows.Forms.RadioButton btnFollowLink;
		private System.Windows.Forms.RadioButton btnResume;

		// event arguments
        private IssueEventArgs _arguments;

		// list of problem types the user chose to skip
		private readonly Hashtable _skipProblemTypes = new Hashtable();

		// true if the user chose to overwrite all existing files
	    private FileOverwriteCondition _overwriteCondition;

		// true if the user chose to overwrite all older files
        private RadioButton btnOverwriteDiffChecksum;

		// true if the user chose to overwrite all files with a different size
		private Button okBtn;
        private Button retryBtn;
        private Button cancelBtn;
        private CheckBox rememberAction;

		public TransferIssueForm()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();
		}

		/// <summary>
		/// Initialize the form for using it again in new batch transfer.
		/// </summary>
		/// <remarks>
		/// Clears the user's chosen default actions.
		/// </remarks>
		public void Initialize()
		{
			_skipProblemTypes.Clear();
		}

		/// <summary>
		/// Shows the form as a modal dialog box.
		/// The form's buttons are shown according to possible actions taken from the given event argument.
		/// </summary>
		/// <param name="e">Event argument describing the type of a problem.</param>
        public void ShowModal(Control form, IssueEventArgs e)
		{
			if (e == null)
				throw new ArgumentNullException("e");

			// if the user chose to skip this problem type, skip it
            if (_skipProblemTypes.ContainsKey(e.IssueType))
			{
                if (e.IssueType == TransferIssueType.FileExists)
                    e.Overwrite(_overwriteCondition);
                else
			        e.Skip();
				return;
			}

			// if the problem is FileExists check, whether the user have chosen any default action already
            if (e.IssueType == TransferIssueType.FileExists)
			{
				// format the text according to action in progress (Downloading or Uploading)
				string messageFormat = "Overwrite file: {0}\n\t{1} Bytes, {2}\n\nwith file: {3}\n\t{4} Bytes, {5}";
				if (e.Action == FileTransferType.Downloading)
				{
					lblMessage.Text = string.Format(messageFormat,
						e.LocalPath, e.LocalItem.Length, e.LocalItem.LastWriteTime,
						e.RemotePath, e.RemoteItem.Length, e.RemoteItem.LastWriteTime);
				}
				else
				{
					lblMessage.Text = string.Format(messageFormat,
						e.RemotePath, e.RemoteItem.Length, e.RemoteItem.LastWriteTime,
						e.LocalPath, e.LocalItem.Length, e.LocalItem.LastWriteTime);
				}

				lblMessage.TextAlign = ContentAlignment.MiddleLeft;
			    this.Text = "Target file already exists";
			}
			else
			{
                lblMessage.Text = e.Exception.Message;
				lblMessage.TextAlign = ContentAlignment.MiddleCenter;
			}

			// store the event arguments for later use at button click handler
			_arguments = e;

			// only enable buttons that represent a possible action
			cancelBtn.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Cancel);
			btnSkip.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Skip);
			retryBtn.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Retry);
			btnRename.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Rename);
			btnOverwrite.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Overwrite);
			btnOverwriteOlder.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Overwrite);
			btnOverwriteDiffSize.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Overwrite);
            btnOverwriteDiffChecksum.Enabled = e.IsOverwriteConditionPossible(FileOverwriteCondition.ChecksumDiffers);
			btnFollowLink.Enabled = e.IsReactionPossible(FileTransferIssueReaction.FollowLink);
			btnResume.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Resume);

			// select default action button
			SetDefaultActionButton(e);

			ShowDialog(form);
		}

		/// <summary>
		/// Selects the button according to default action.
		/// </summary>
		/// <param name="action">Default action.</param>
        private void SetDefaultActionButton(IssueEventArgs e)
		{
			switch (e.Reaction)
			{
				case FileTransferIssueReaction.Cancel: cancelBtn.Select(); break;
				case FileTransferIssueReaction.Fail: cancelBtn.Select(); break;
				case FileTransferIssueReaction.Skip: btnSkip.Select(); break;
				case FileTransferIssueReaction.Retry: retryBtn.Select(); break;
				case FileTransferIssueReaction.Rename: btnRename.Select(); break;
				case FileTransferIssueReaction.FollowLink: btnFollowLink.Select(); break;
				case FileTransferIssueReaction.Resume: btnResume.Select(); break;
				case FileTransferIssueReaction.Overwrite:
					switch (e.OverwriteCondition)
					{
						case FileOverwriteCondition.None: btnOverwrite.Select(); break;
                        case FileOverwriteCondition.Older: btnOverwriteOlder.Select(); break;
                        case FileOverwriteCondition.SizeDiffers: btnOverwriteDiffSize.Select(); break;
                        case FileOverwriteCondition.ChecksumDiffers: btnOverwriteDiffChecksum.Select(); break;
					}
					break;
			}
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose(bool disposing)
		{
			if (disposing)
			{
				if (components != null)
				{
					components.Dispose();
				}
			}
			base.Dispose(disposing);
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            this.btnSkip = new System.Windows.Forms.RadioButton();
            this.btnRename = new System.Windows.Forms.RadioButton();
            this.btnOverwrite = new System.Windows.Forms.RadioButton();
            this.btnOverwriteOlder = new System.Windows.Forms.RadioButton();
            this.btnOverwriteDiffSize = new System.Windows.Forms.RadioButton();
            this.btnFollowLink = new System.Windows.Forms.RadioButton();
            this.lblMessage = new System.Windows.Forms.Label();
            this.btnResume = new System.Windows.Forms.RadioButton();
            this.btnOverwriteDiffChecksum = new System.Windows.Forms.RadioButton();
            this.okBtn = new System.Windows.Forms.Button();
            this.retryBtn = new System.Windows.Forms.Button();
            this.cancelBtn = new System.Windows.Forms.Button();
            this.rememberAction = new System.Windows.Forms.CheckBox();
            this.SuspendLayout();
            // 
            // btnSkip
            // 
            this.btnSkip.Location = new System.Drawing.Point(271, 160);
            this.btnSkip.Name = "btnSkip";
            this.btnSkip.Size = new System.Drawing.Size(179, 24);
            this.btnSkip.TabIndex = 7;
            this.btnSkip.Text = "&Skip";
            this.btnSkip.Click += new System.EventHandler(this.btnSkip_Click);
            // 
            // btnRename
            // 
            this.btnRename.Location = new System.Drawing.Point(271, 130);
            this.btnRename.Name = "btnRename";
            this.btnRename.Size = new System.Drawing.Size(157, 24);
            this.btnRename.TabIndex = 6;
            this.btnRename.Text = "&Rename...";
            this.btnRename.Click += new System.EventHandler(this.btnRename_Click);
            // 
            // btnOverwrite
            // 
            this.btnOverwrite.Location = new System.Drawing.Point(12, 100);
            this.btnOverwrite.Name = "btnOverwrite";
            this.btnOverwrite.Size = new System.Drawing.Size(157, 24);
            this.btnOverwrite.TabIndex = 1;
            this.btnOverwrite.Text = "&Overwrite";
            // 
            // btnOverwriteOlder
            // 
            this.btnOverwriteOlder.Location = new System.Drawing.Point(12, 130);
            this.btnOverwriteOlder.Name = "btnOverwriteOlder";
            this.btnOverwriteOlder.Size = new System.Drawing.Size(157, 24);
            this.btnOverwriteOlder.TabIndex = 2;
            this.btnOverwriteOlder.Text = "Overwrite Older File";
            // 
            // btnOverwriteDiffSize
            // 
            this.btnOverwriteDiffSize.Location = new System.Drawing.Point(11, 160);
            this.btnOverwriteDiffSize.Name = "btnOverwriteDiffSize";
            this.btnOverwriteDiffSize.Size = new System.Drawing.Size(157, 24);
            this.btnOverwriteDiffSize.TabIndex = 3;
            this.btnOverwriteDiffSize.Text = "Overwrite &If Size Differs";
            // 
            // btnFollowLink
            // 
            this.btnFollowLink.Location = new System.Drawing.Point(271, 190);
            this.btnFollowLink.Name = "btnFollowLink";
            this.btnFollowLink.Size = new System.Drawing.Size(157, 24);
            this.btnFollowLink.TabIndex = 8;
            this.btnFollowLink.Text = "Follow &Link";
            // 
            // lblMessage
            // 
            this.lblMessage.Anchor = ((System.Windows.Forms.AnchorStyles)(((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Left)
                        | System.Windows.Forms.AnchorStyles.Right)));
            this.lblMessage.Location = new System.Drawing.Point(8, 8);
            this.lblMessage.Name = "lblMessage";
            this.lblMessage.Size = new System.Drawing.Size(580, 89);
            this.lblMessage.TabIndex = 11;
            this.lblMessage.Text = "Message";
            this.lblMessage.TextAlign = System.Drawing.ContentAlignment.MiddleCenter;
            // 
            // btnResume
            // 
            this.btnResume.Location = new System.Drawing.Point(271, 100);
            this.btnResume.Name = "btnResume";
            this.btnResume.Size = new System.Drawing.Size(179, 24);
            this.btnResume.TabIndex = 5;
            this.btnResume.Text = "Resume";
            // 
            // btnOverwriteDiffChecksum
            // 
            this.btnOverwriteDiffChecksum.Location = new System.Drawing.Point(11, 190);
            this.btnOverwriteDiffChecksum.Name = "btnOverwriteDiffChecksum";
            this.btnOverwriteDiffChecksum.Size = new System.Drawing.Size(179, 24);
            this.btnOverwriteDiffChecksum.TabIndex = 4;
            this.btnOverwriteDiffChecksum.Text = "Overwrite If Checksum Differs";
            // 
            // okBtn
            // 
            this.okBtn.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Right)));
            this.okBtn.Location = new System.Drawing.Point(349, 264);
            this.okBtn.Name = "okBtn";
            this.okBtn.Size = new System.Drawing.Size(75, 23);
            this.okBtn.TabIndex = 31;
            this.okBtn.Text = "OK";
            this.okBtn.UseVisualStyleBackColor = true;
            this.okBtn.Click += new System.EventHandler(this.okBtn_Click);
            // 
            // retryBtn
            // 
            this.retryBtn.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Right)));
            this.retryBtn.Location = new System.Drawing.Point(430, 264);
            this.retryBtn.Name = "retryBtn";
            this.retryBtn.Size = new System.Drawing.Size(75, 23);
            this.retryBtn.TabIndex = 32;
            this.retryBtn.Text = "Retry";
            this.retryBtn.UseVisualStyleBackColor = true;
            this.retryBtn.Click += new System.EventHandler(this.btnRetry_Click);
            // 
            // cancelBtn
            // 
            this.cancelBtn.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Bottom | System.Windows.Forms.AnchorStyles.Right)));
            this.cancelBtn.Location = new System.Drawing.Point(511, 264);
            this.cancelBtn.Name = "cancelBtn";
            this.cancelBtn.Size = new System.Drawing.Size(75, 23);
            this.cancelBtn.TabIndex = 33;
            this.cancelBtn.Text = "Cancel";
            this.cancelBtn.UseVisualStyleBackColor = true;
            this.cancelBtn.Click += new System.EventHandler(this.btnCancel_Click);
            // 
            // rememberAction
            // 
            this.rememberAction.AutoSize = true;
            this.rememberAction.Location = new System.Drawing.Point(11, 230);
            this.rememberAction.Name = "rememberAction";
            this.rememberAction.Size = new System.Drawing.Size(141, 17);
            this.rememberAction.TabIndex = 20;
            this.rememberAction.Text = "Use this action next time";
            this.rememberAction.UseVisualStyleBackColor = true;
            // 
            // TransferIssueForm
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(598, 299);
            this.Controls.Add(this.rememberAction);
            this.Controls.Add(this.cancelBtn);
            this.Controls.Add(this.retryBtn);
            this.Controls.Add(this.okBtn);
            this.Controls.Add(this.btnOverwrite);
            this.Controls.Add(this.btnOverwriteOlder);
            this.Controls.Add(this.btnOverwriteDiffChecksum);
            this.Controls.Add(this.btnResume);
            this.Controls.Add(this.lblMessage);
            this.Controls.Add(this.btnFollowLink);
            this.Controls.Add(this.btnOverwriteDiffSize);
            this.Controls.Add(this.btnRename);
            this.Controls.Add(this.btnSkip);
            this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog;
            this.MaximizeBox = false;
            this.MinimizeBox = false;
            this.Name = "TransferIssueForm";
            this.ShowInTaskbar = false;
            this.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent;
            this.Text = "Transfer Problem Detected";
            this.ResumeLayout(false);
            this.PerformLayout();

		}
		#endregion

		private void btnCancel_Click(object sender, System.EventArgs e)
		{
			_arguments.Cancel();
			Close();
		}

		private void btnSkip_Click(object sender, System.EventArgs e)
		{
			_arguments.Skip();
			Close();
		}

		private void btnRetry_Click(object sender, System.EventArgs e)
		{
			_arguments.Retry();
			Close();
		}

		private void btnRename_Click(object sender, System.EventArgs e)
		{
			// initialize the renaming form
			NewNameForm formNewName = new NewNameForm();

			// set the current file name
			string oldName = Path.GetFileName(_arguments.LocalPath);
			formNewName.NewName = oldName;

			// show the form
			DialogResult result = formNewName.ShowDialog(this);

			// get the new name
			string newName = formNewName.NewName;

			// check whether the user clicked on OK and insert something nonempty and something else
			if (result != DialogResult.OK || newName.Length == 0 || newName == oldName)
				return;

			// set the appropriate action and new name to the event arguments
			_arguments.Rename(newName);
			Close();
		}

		void DoAction()
        {
            if (btnSkip.Checked)
                _arguments.Skip();
            else if (btnOverwrite.Checked)
            {
                _arguments.Overwrite();
                _overwriteCondition = FileOverwriteCondition.None;
            }
            else if (btnOverwriteOlder.Checked)
            {
                _arguments.Overwrite(FileOverwriteCondition.Older);
                _overwriteCondition = FileOverwriteCondition.Older;
            }
            else if (btnOverwriteDiffSize.Checked)
            {
                _arguments.Overwrite(FileOverwriteCondition.SizeDiffers);
                _overwriteCondition = FileOverwriteCondition.SizeDiffers;
            }
            else if (btnOverwriteDiffChecksum.Checked)
            {
                _arguments.Overwrite(FileOverwriteCondition.ChecksumDiffers);
                _overwriteCondition = FileOverwriteCondition.ChecksumDiffers;
            }
            else if (btnFollowLink.Checked)
                _arguments.FollowLink();
            else if (btnResume.Checked)
                _arguments.Resume();

            Close();
        }

        private void okBtn_Click(object sender, EventArgs e)
        {
            if (rememberAction.Checked)
                // add the problem to the table of automatically skipped problems
                _skipProblemTypes.Add(_arguments.IssueType, null);

            DoAction();
        }
	}
}

File: Examples/FxSocket.Sftp/Windows/VB/SftpDesktopClientDemo/TransferIssueForm.vb
Imports System.IO
Imports System.Collections
Imports System.ComponentModel
Imports FxSocket.IO
Imports FxSocket.FileSystem
Imports FxSocket.Net

Namespace FxSocketSamples
	''' <summary>
	''' Problem handling form.
	''' </summary>
	Public Class TransferIssueForm
		Inherits System.Windows.Forms.Form
		''' <summary>
		''' Required designer variable.
		''' </summary>
		Private components As System.ComponentModel.Container = Nothing

		Private lblMessage As System.Windows.Forms.Label
		Private WithEvents btnSkip As System.Windows.Forms.RadioButton
		Private WithEvents btnRename As System.Windows.Forms.RadioButton
		Private btnOverwrite As System.Windows.Forms.RadioButton
		Private btnOverwriteOlder As System.Windows.Forms.RadioButton
		Private btnOverwriteDiffSize As System.Windows.Forms.RadioButton
		Private btnFollowLink As System.Windows.Forms.RadioButton
		Private btnResume As System.Windows.Forms.RadioButton

		' event arguments
		Private _arguments As IssueEventArgs

		' list of problem types the user chose to skip
		Private ReadOnly _skipProblemTypes As New Hashtable()

		' true if the user chose to overwrite all existing files
		Private _overwriteCondition As FileOverwriteCondition

		' true if the user chose to overwrite all older files
		Private btnOverwriteDiffChecksum As RadioButton

		' true if the user chose to overwrite all files with a different size
		Private WithEvents okBtn As Button
		Private WithEvents retryBtn As Button
		Private WithEvents cancelBtn As Button
		Private rememberAction As CheckBox

		Public Sub New()
			'
			' Required for Windows Form Designer support
			'
			InitializeComponent()
		End Sub

		''' <summary>
		''' Initialize the form for using it again in new batch transfer.
		''' </summary>
		''' <remarks>
		''' Clears the user's chosen default actions.
		''' </remarks>
		Public Sub Initialize()
			_skipProblemTypes.Clear()
		End Sub

		''' <summary>
		''' Shows the form as a modal dialog box.
		''' The form's buttons are shown according to possible actions taken from the given event argument.
		''' </summary>
		''' <param name="e">Event argument describing the type of a problem.</param>
		Public Sub ShowModal(ByVal form As Control, ByVal e As IssueEventArgs)
			If e Is Nothing Then
				Throw New ArgumentNullException("e")
			End If

			' if the user chose to skip this problem type, skip it
			If _skipProblemTypes.ContainsKey(e.IssueType) Then
				If e.IssueType = TransferIssueType.FileExists Then
					e.Overwrite(_overwriteCondition)
				Else
					e.Skip()
				End If
				Return
			End If

			' if the problem is FileExists check, whether the user have chosen any default action already
			If e.IssueType = TransferIssueType.FileExists Then
				' format the text according to action in progress (Downloading or Uploading)
				Dim messageFormat As String = "Overwrite file: {0}" & vbLf & vbTab & "{1} Bytes, {2}" & vbLf & vbLf & "with file: {3}" & vbLf & vbTab & "{4} Bytes, {5}"
				If e.Action = FileTransferType.Downloading Then
					lblMessage.Text = String.Format(messageFormat, e.LocalPath, e.LocalItem.Length, e.LocalItem.LastWriteTime, e.RemotePath, e.RemoteItem.Length, e.RemoteItem.LastWriteTime)
				Else
					lblMessage.Text = String.Format(messageFormat, e.RemotePath, e.RemoteItem.Length, e.RemoteItem.LastWriteTime, e.LocalPath, e.LocalItem.Length, e.LocalItem.LastWriteTime)
				End If

				lblMessage.TextAlign = ContentAlignment.MiddleLeft
				Me.Text = "Target file already exists"
			Else
				lblMessage.Text = e.Exception.Message
				lblMessage.TextAlign = ContentAlignment.MiddleCenter
			End If

			' store the event arguments for later use at button click handler
			_arguments = e

			' only enable buttons that represent a possible action
			cancelBtn.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Cancel)
			btnSkip.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Skip)
			retryBtn.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Retry)
			btnRename.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Rename)
			btnOverwrite.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Overwrite)
			btnOverwriteOlder.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Overwrite)
			btnOverwriteDiffSize.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Overwrite)
			btnOverwriteDiffChecksum.Enabled = e.IsOverwriteConditionPossible(FileOverwriteCondition.ChecksumDiffers)
			btnFollowLink.Enabled = e.IsReactionPossible(FileTransferIssueReaction.FollowLink)
			btnResume.Enabled = e.IsReactionPossible(FileTransferIssueReaction.Resume)

			' select default action button
			SetDefaultActionButton(e)

			ShowDialog(form)
		End Sub

		''' <summary>
		''' Selects the button according to default action.
		''' </summary>
		''' <param name="action">Default action.</param>
		Private Sub SetDefaultActionButton(ByVal e As IssueEventArgs)
			Select Case e.Reaction
				Case FileTransferIssueReaction.Cancel
					cancelBtn.Select()
				Case FileTransferIssueReaction.Fail
					cancelBtn.Select()
				Case FileTransferIssueReaction.Skip
					btnSkip.Select()
				Case FileTransferIssueReaction.Retry
					retryBtn.Select()
				Case FileTransferIssueReaction.Rename
					btnRename.Select()
				Case FileTransferIssueReaction.FollowLink
					btnFollowLink.Select()
				Case FileTransferIssueReaction.Resume
					btnResume.Select()
				Case FileTransferIssueReaction.Overwrite
					Select Case e.OverwriteCondition
						Case FileOverwriteCondition.None
							btnOverwrite.Select()
						Case FileOverwriteCondition.Older
							btnOverwriteOlder.Select()
						Case FileOverwriteCondition.SizeDiffers
							btnOverwriteDiffSize.Select()
						Case FileOverwriteCondition.ChecksumDiffers
							btnOverwriteDiffChecksum.Select()
					End Select
			End Select
		End Sub

		''' <summary>
		''' Clean up any resources being used.
		''' </summary>
		Protected Overrides Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If components IsNot Nothing Then
					components.Dispose()
				End If
			End If
			MyBase.Dispose(disposing)
		End Sub

		#Region "Windows Form Designer generated code"
		''' <summary>
		''' Required method for Designer support - do not modify
		''' the contents of this method with the code editor.
		''' </summary>
		Private Sub InitializeComponent()
			Me.btnSkip = New System.Windows.Forms.RadioButton()
			Me.btnRename = New System.Windows.Forms.RadioButton()
			Me.btnOverwrite = New System.Windows.Forms.RadioButton()
			Me.btnOverwriteOlder = New System.Windows.Forms.RadioButton()
			Me.btnOverwriteDiffSize = New System.Windows.Forms.RadioButton()
			Me.btnFollowLink = New System.Windows.Forms.RadioButton()
			Me.lblMessage = New System.Windows.Forms.Label()
			Me.btnResume = New System.Windows.Forms.RadioButton()
			Me.btnOverwriteDiffChecksum = New System.Windows.Forms.RadioButton()
			Me.okBtn = New System.Windows.Forms.Button()
			Me.retryBtn = New System.Windows.Forms.Button()
			Me.cancelBtn = New System.Windows.Forms.Button()
			Me.rememberAction = New System.Windows.Forms.CheckBox()
			Me.SuspendLayout()
			' 
			' btnSkip
			' 
			Me.btnSkip.Location = New System.Drawing.Point(271, 160)
			Me.btnSkip.Name = "btnSkip"
			Me.btnSkip.Size = New System.Drawing.Size(179, 24)
			Me.btnSkip.TabIndex = 7
			Me.btnSkip.Text = "&Skip"
'			Me.btnSkip.Click += New System.EventHandler(Me.btnSkip_Click)
			' 
			' btnRename
			' 
			Me.btnRename.Location = New System.Drawing.Point(271, 130)
			Me.btnRename.Name = "btnRename"
			Me.btnRename.Size = New System.Drawing.Size(157, 24)
			Me.btnRename.TabIndex = 6
			Me.btnRename.Text = "&Rename..."
'			Me.btnRename.Click += New System.EventHandler(Me.btnRename_Click)
			' 
			' btnOverwrite
			' 
			Me.btnOverwrite.Location = New System.Drawing.Point(12, 100)
			Me.btnOverwrite.Name = "btnOverwrite"
			Me.btnOverwrite.Size = New System.Drawing.Size(157, 24)
			Me.btnOverwrite.TabIndex = 1
			Me.btnOverwrite.Text = "&Overwrite"
			' 
			' btnOverwriteOlder
			' 
			Me.btnOverwriteOlder.Location = New System.Drawing.Point(12, 130)
			Me.btnOverwriteOlder.Name = "btnOverwriteOlder"
			Me.btnOverwriteOlder.Size = New System.Drawing.Size(157, 24)
			Me.btnOverwriteOlder.TabIndex = 2
			Me.btnOverwriteOlder.Text = "Overwrite Older File"
			' 
			' btnOverwriteDiffSize
			' 
			Me.btnOverwriteDiffSize.Location = New System.Drawing.Point(11, 160)
			Me.btnOverwriteDiffSize.Name = "btnOverwriteDiffSize"
			Me.btnOverwriteDiffSize.Size = New System.Drawing.Size(157, 24)
			Me.btnOverwriteDiffSize.TabIndex = 3
			Me.btnOverwriteDiffSize.Text = "Overwrite &If Size Differs"
			' 
			' btnFollowLink
			' 
			Me.btnFollowLink.Location = New System.Drawing.Point(271, 190)
			Me.btnFollowLink.Name = "btnFollowLink"
			Me.btnFollowLink.Size = New System.Drawing.Size(157, 24)
			Me.btnFollowLink.TabIndex = 8
			Me.btnFollowLink.Text = "Follow &Link"
			' 
			' lblMessage
			' 
			Me.lblMessage.Anchor = (CType(((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Left) Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles))
			Me.lblMessage.Location = New System.Drawing.Point(8, 8)
			Me.lblMessage.Name = "lblMessage"
			Me.lblMessage.Size = New System.Drawing.Size(580, 89)
			Me.lblMessage.TabIndex = 11
			Me.lblMessage.Text = "Message"
			Me.lblMessage.TextAlign = System.Drawing.ContentAlignment.MiddleCenter
			' 
			' btnResume
			' 
			Me.btnResume.Location = New System.Drawing.Point(271, 100)
			Me.btnResume.Name = "btnResume"
			Me.btnResume.Size = New System.Drawing.Size(179, 24)
			Me.btnResume.TabIndex = 5
			Me.btnResume.Text = "Resume"
			' 
			' btnOverwriteDiffChecksum
			' 
			Me.btnOverwriteDiffChecksum.Location = New System.Drawing.Point(11, 190)
			Me.btnOverwriteDiffChecksum.Name = "btnOverwriteDiffChecksum"
			Me.btnOverwriteDiffChecksum.Size = New System.Drawing.Size(179, 24)
			Me.btnOverwriteDiffChecksum.TabIndex = 4
			Me.btnOverwriteDiffChecksum.Text = "Overwrite If Checksum Differs"
			' 
			' okBtn
			' 
			Me.okBtn.Anchor = (CType((System.Windows.Forms.AnchorStyles.Bottom Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles))
			Me.okBtn.Location = New System.Drawing.Point(349, 264)
			Me.okBtn.Name = "okBtn"
			Me.okBtn.Size = New System.Drawing.Size(75, 23)
			Me.okBtn.TabIndex = 31
			Me.okBtn.Text = "OK"
			Me.okBtn.UseVisualStyleBackColor = True
'			Me.okBtn.Click += New System.EventHandler(Me.okBtn_Click)
			' 
			' retryBtn
			' 
			Me.retryBtn.Anchor = (CType((System.Windows.Forms.AnchorStyles.Bottom Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles))
			Me.retryBtn.Location = New System.Drawing.Point(430, 264)
			Me.retryBtn.Name = "retryBtn"
			Me.retryBtn.Size = New System.Drawing.Size(75, 23)
			Me.retryBtn.TabIndex = 32
			Me.retryBtn.Text = "Retry"
			Me.retryBtn.UseVisualStyleBackColor = True
'			Me.retryBtn.Click += New System.EventHandler(Me.btnRetry_Click)
			' 
			' cancelBtn
			' 
			Me.cancelBtn.Anchor = (CType((System.Windows.Forms.AnchorStyles.Bottom Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles))
			Me.cancelBtn.Location = New System.Drawing.Point(511, 264)
			Me.cancelBtn.Name = "cancelBtn"
			Me.cancelBtn.Size = New System.Drawing.Size(75, 23)
			Me.cancelBtn.TabIndex = 33
			Me.cancelBtn.Text = "Cancel"
			Me.cancelBtn.UseVisualStyleBackColor = True
'			Me.cancelBtn.Click += New System.EventHandler(Me.btnCancel_Click)
			' 
			' rememberAction
			' 
			Me.rememberAction.AutoSize = True
			Me.rememberAction.Location = New System.Drawing.Point(11, 230)
			Me.rememberAction.Name = "rememberAction"
			Me.rememberAction.Size = New System.Drawing.Size(141, 17)
			Me.rememberAction.TabIndex = 20
			Me.rememberAction.Text = "Use this action next time"
			Me.rememberAction.UseVisualStyleBackColor = True
			' 
			' TransferIssueForm
			' 
			Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
			Me.ClientSize = New System.Drawing.Size(598, 299)
			Me.Controls.Add(Me.rememberAction)
			Me.Controls.Add(Me.cancelBtn)
			Me.Controls.Add(Me.retryBtn)
			Me.Controls.Add(Me.okBtn)
			Me.Controls.Add(Me.btnOverwrite)
			Me.Controls.Add(Me.btnOverwriteOlder)
			Me.Controls.Add(Me.btnOverwriteDiffChecksum)
			Me.Controls.Add(Me.btnResume)
			Me.Controls.Add(Me.lblMessage)
			Me.Controls.Add(Me.btnFollowLink)
			Me.Controls.Add(Me.btnOverwriteDiffSize)
			Me.Controls.Add(Me.btnRename)
			Me.Controls.Add(Me.btnSkip)
			Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedDialog
			Me.MaximizeBox = False
			Me.MinimizeBox = False
			Me.Name = "TransferIssueForm"
			Me.ShowInTaskbar = False
			Me.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent
			Me.Text = "Transfer Problem Detected"
			Me.ResumeLayout(False)
			Me.PerformLayout()

		End Sub
		#End Region

		Private Sub btnCancel_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles cancelBtn.Click
			_arguments.Cancel()
			Close()
		End Sub

		Private Sub btnSkip_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnSkip.Click
			_arguments.Skip()
			Close()
		End Sub

		Private Sub btnRetry_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles retryBtn.Click
			_arguments.Retry()
			Close()
		End Sub

		Private Sub btnRename_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnRename.Click
			' initialize the renaming form
			Dim formNewName As New NewNameForm()

			' set the current file name
			Dim oldName As String = Path.GetFileName(_arguments.LocalPath)
			formNewName.NewName = oldName

			' show the form
			Dim result As DialogResult = formNewName.ShowDialog(Me)

			' get the new name
			Dim newName As String = formNewName.NewName

			' check whether the user clicked on OK and insert something nonempty and something else
			If result <> System.Windows.Forms.DialogResult.OK OrElse newName.Length = 0 OrElse newName = oldName Then
				Return
			End If

			' set the appropriate action and new name to the event arguments
			_arguments.Rename(newName)
			Close()
		End Sub

		Private Sub DoAction()
			If btnSkip.Checked Then
				_arguments.Skip()
			ElseIf btnOverwrite.Checked Then
				_arguments.Overwrite()
				_overwriteCondition = FileOverwriteCondition.None
			ElseIf btnOverwriteOlder.Checked Then
				_arguments.Overwrite(FileOverwriteCondition.Older)
				_overwriteCondition = FileOverwriteCondition.Older
			ElseIf btnOverwriteDiffSize.Checked Then
				_arguments.Overwrite(FileOverwriteCondition.SizeDiffers)
				_overwriteCondition = FileOverwriteCondition.SizeDiffers
			ElseIf btnOverwriteDiffChecksum.Checked Then
				_arguments.Overwrite(FileOverwriteCondition.ChecksumDiffers)
				_overwriteCondition = FileOverwriteCondition.ChecksumDiffers
			ElseIf btnFollowLink.Checked Then
				_arguments.FollowLink()
			ElseIf btnResume.Checked Then
				_arguments.Resume()
			End If

			Close()
		End Sub

		Private Sub okBtn_Click(ByVal sender As Object, ByVal e As EventArgs) Handles okBtn.Click
			If rememberAction.Checked Then
				' add the problem to the table of automatically skipped problems
				_skipProblemTypes.Add(_arguments.IssueType, Nothing)
			End If

			DoAction()
		End Sub
	End Class
End Namespace
